source/userguide/text_preparation_pipeline.rst

```diff
@@ -0,0 +1,661 @@
+This section includes boilerplate code to build a custom normalizer class. Let's recreate the :attr:`Lowercase` normalizer class.
+A custom tokenizer must extend from MindMeld's abstract :attr:`Normalizer` class:
+
+
+.. code:: python
+
+    from abc import ABC, abstractmethod
+
+    class Normalizer(ABC):
+        """Abstract Normalizer Base Class."""
+
+        @abstractmethod
+        def normalize(self, text):
+            """
+            Args:
+                text (str): Text to normalize.
+            Returns:
+                normalized_text (str): Normalized Text.
+            """
+            raise NotImplementedError("Subclasses must implement this method")
+
+
+With this in mind, let's recreate MindMeld's :attr:`Lowercase` normalizer class.
+
+.. code:: python
+
+    from mindmeld.text_preparation.normalizers import Normalizer
+
+    class Lowercase(Normalizer):
+
+        def normalize(self, text):
+            return text.lower()
+
+
+This normalizer would transform the text "I Like to Run!" to "i like to run!".
+The steps to use a custom Normalizer in your application are explained here.
+
+
+.. note::
+
+    MindMeld normalizes queries on a per-token basis. Custom normalizers should be designed to normalize individual tokens and not sentences as a whole.
+
+
+Stemming
+--------
+Stemming is the process of reducing a word to its stem or root. If a stemmer is not specified in the :attr:`TEXT_PREPARATION_CONFIG`, then MindMeld will automatically select a stemmer
+based on the language of the application.
+
+
+EnglishNLTKStemmer
+^^^^^^^^^^^^^^^^^^
+
+The :attr:`EnglishNLTKStemmer` stemmer uses the :attr:`PorterStemmer` from the nltk library. The porter stemmer implements a series of rules that determine common suffixes from sentences.
+This includes removing the letter "s" from plural words or "ing" from gerunds.
+
+Let's take a look at a few examples of the :attr:`EnglishNLTKStemmer`. First we'll make an instance of the stemmer:
+
+.. code:: python
+
+    from mindmeld.text_preparation.stemmers import EnglishNLTKStemmer
+    english_nltk_stemmer = EnglishNLTKStemmer()
+
+Now let's stem the words "running" and "pearls".
+
+.. code:: python
+
+    >>> print(english_nltk_stemmer.stem_word("running"))
+    >>> run
+    >>> print(english_nltk_stemmer.stem_word("pearls"))
+    >>> pearl
+
+As expected, the stemmer removes "ing" from "running" and the "s" from "pearls" to create stemmed words.
+
+
+SnowballNLTKStemmer
+^^^^^^^^^^^^^^^^^^^
+
+The :attr:`SnowballNLTKStemmer` stemmer works in a similar manner to the :attr:`EnglishNLTKStemmer`, however, it offers support for a larger set of languages.
+Namely, the :attr:`SnowballNLTKStemmer` supports Danish (da), Dutch (nl), Finnish (fi), French (fr), German (de), Hungarian (hu), Italian (it), Norwegian (nb), Portuguese (pt), Romanian (ro), Russian (ru), Spanish (es) and Swedish (sv).
+
+To create an instance of the :attr:`SnowballNLTKStemmer`, we can use MindMeld's :attr:`StemmerFactory`.
```

Comment:
Why is this better than directly calling 'SnowballNLTKStemmer'? (I suspect the answer is that the stemmer factory lets you call by language code, where as the stemmer class expects a full language name, but is there more to it?) Notably, the full Snowball/Porter stemmer with derivational suffixes removed is also accessible through `SnowballNLTKStemmer("english")`.

Reply:
Yes, that's correct. The stemmer factory lets you call by language code and converts that code into a lower-cased version of the language like "spanish" for exampe. Aside from this, there isn't any other main difference. A user could alternatively do SnowballNLTKStemmer("spanish"), for example. Should I switch the example to use the latter?